src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -0,0 +1,1370 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+import ast
+import datetime
+import decimal
+import inspect
+import logging
+import os
+import platform
+import sys
+import typing
+from functools import reduce
+from logging import getLogger
+from pathlib import Path
+from types import ModuleType
+from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Tuple, Union
+
+import dateutil
+from dateutil.tz import tzlocal
+
+import snowflake.snowpark
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark._internal.analyzer.expression import (
+    Attribute,
+    CaseWhen,
+    Expression,
+    FunctionExpression,
+    Literal,
+    MultipleExpression,
+    Star,
+    UnresolvedAttribute,
+)
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import SaveMode
+from snowflake.snowpark._internal.analyzer.unary_expression import Alias
+from snowflake.snowpark._internal.ast import AstBatch
+from snowflake.snowpark._internal.type_utils import (
+    VALID_PYTHON_TYPES_FOR_LITERAL_VALUE,
+    ColumnOrLiteral,
+    ColumnOrName,
+    ColumnOrSqlExpr,
+)
+from snowflake.snowpark._internal.utils import str_to_enum
+from snowflake.snowpark.types import DataType, StructType
+
+# This flag causes an explicit error to be raised if any Snowpark object instance is missing an AST or field, when this
+# AST or field is required to populate the AST field of a different Snowpark object instance.
+FAIL_ON_MISSING_AST = True
+
+# The path to the snowpark package.
+SNOWPARK_LIB_PATH = Path(__file__).parent.parent.resolve()
+
+# Test mode. In test mode, the source filename is ignored.
+SRC_POSITION_TEST_MODE = False
+
+_logger = getLogger(__name__)
+
+
+# TODO(SNOW-1491199) - This method is not covered by tests until the end of phase 0. Drop the pragma when it is covered.
```

Comment:
Since all code here are not used and tested, why we include it here? We could include those utility file when it is used in the future PRs.

The reason I suggest removing it here is that
1. It is hard to review all of them without context and tests;
2. If we just let it in, later we may forget to review them.

Reply:
We discussed this when splitting the changes. To see how all code is used together, we can look at the staging branch, https://github.com/snowflakedb/snowpark-python/tree/ls-SNOW-1491199-merge-phase0-server-side

Instead of pursuing a single, huge PR (https://github.com/snowflakedb/snowpark-python/pull/2136), our finer-grained choices are to merge at roughly module level, in a bottom-up fashion (this approach), or at an even finer-grained level, "vertically" - add a set of features and all internal functionality that they depend on.

Splitting the top level change at the finest-grained level is a lot more work than organizing modules in a bottom-up fashion. The context is available in https://github.com/snowflakedb/snowpark-python/pull/2136